### ISSUE TITLE
Error : Unsupported config option for services. 'enableMQTT'

### ISSUE THREAD

**Issue by sk0010637** [ptarmiganlabs/butler-sos#120]
Hi, I am trying to add new container in docker-compose.yml.  However getting error when trying to restart the docker container
as Error : Unsupported config option for services. 'enableMQTT'.

version: '3.3'

Can you please suggest.

**Comment by mountaindude** [comment]
Hi there,

First, version 3.3 of Butler SOS is very old. I'm not even sure that version was ever released..
Released versions are listed here: https://github.com/ptarmiganlabs/butler-sos/releases
Docker images listed here: https://hub.docker.com/r/ptarmiganlabs/butler-sos/tags


Then I need more info to help you. For starters:

- A copy of your YAML/JSON config file (make sure to mask out any sensitive info in it before posting it here). Impossible to tell what's going on without access to that file.
- Info about what platform you use. Linux, Windows or something else? Version?
- Node.js version?

**Comment by sk0010637** [comment]
Hello,

Many thanks for the reply.

I am attaching the docker-compose.yml and production.yaml files which are used.

production.yaml is the file which controls what and how DEV/PROD servers are monitored.
Have attached the .txt version of the files.

I am trying to add one more container (butler-sos_FIN_dev) in docker-compose.yml file  
While trying to restart this container, I am getting "Unsupported config option for services. 'enableMQTT'" 
Platform used - Linux

Please let me know if anymore details are required.

**Comment by sk0010637** [comment]
I am unable to attach the files.

**Comment by mountaindude** [comment]
Copy paste the files,  maybe?

**Comment by sk0010637** [comment]
many thanks again...  Tried all option to attach, copy/paste or drag n drop..  however nothing is working.

If you don't mind.. can we have a quick call to sort this out please...   I would share the webex link..  Please let me know..

**Comment by sk0010637** [comment]
Please find the code as below for the new container which I am trying to add in dockerr-compose.yml file


butler-sos_FIN_dev:
    image: ptarmiganlabs/butler-sos:v4.0.0
    depends_on:
      - influxdb
      - grafana
    container_name: butler-sos_FIN_dev
    restart: always
    volumes:
      # Make config file accessible outside of container
      - "./config_butler-sos_FIN_dev:/nodeapp/config"
      - "./certificate_FIN_dev:/nodeapp/config/certificate"
      - "./logs_butler-sos_dev:/nodeapp/logs"
    environment:
      - "NODE_ENV=production"
      - "NO_PROXY=*"
    networks:
      - senseops
    extra_hosts:
    logging:
      driver: json-file
    user: "504"
    mqttConfig:
    enableMQTT: false



===========================  InfluxDB, grafana



  influxdb:
    image: influxdb:latest
    container_name: influxdb
    restart: always
#    ports:
#      - "8086:8086"
    volumes: 
      - "./influxdb/datastore:/var/lib/influxdb"
    environment:
      - "INFLUXDB_REPORTING_DISABLED=true"
      - "INFLUXDB_BIND_ADDRESS=0.0.0.0:8888"
    networks:
      - senseops
    logging:
      driver: json-file
    user: "504:504"

  grafana:
    image: grafana/grafana:latest
    container_name: grafana
    restart: always
    ports:
      - "3000:3000"
    volumes:
      - "./grafana/datastore:/var/lib/grafana"
    environment:
      - "GF_LOG_LEVEL=WARN"
      - "GF_SERVER_ROOT_URL=http://grafana-qliksense.dev.teliacompany.net:3000"
      - "GF_USERS_ALLOW_SIGN_UP=FALSE"
      - "GF_ANALYTICS_REPORTING_ENABLED=FALSE"
#      - "GF_SMTP_ENABLED=TRUE"
#      - "GF_SMTP_HOST=mail-se-relay.telia.net:25"
#      - "GF_SMTP_HOST=mail.telia.se:25"
#      - "GF_SMTP_SKIP_VERIFY=TRUE"
#      - "GF_SMTP_FROM_ADDRESS=qlik-servicedesk@sec-post.teliasonera.com"
#      - "GF_SMTP_FROM_NAME=SenseOps.noreply"
    networks:
      - senseops
    logging:
      driver: json-file
    user: "504"

  telegraf:
    image: telegraf:latest
    container_name: telegraf
    # links:
    #   - influxdb
    restart: always
    volumes:
      - ./telegraf/telegraf.conf:/etc/telegraf/telegraf.conf:ro
      - /var/run/docker.sock:/var/run/docker.sock:ro
    depends_on:
      - influxdb
    environment: 
      - "REPORTING_DISABLED=true"
    networks:
      - senseops

  chronograf:
    image: chronograf:latest
    container_name: chronograf
    restart: always
    volumes:
      # Mount for chronograf database
      - ./chronograf/data/:/var/lib/chronograf/
    ports:
      # The WebUI for Chronograf is served on port 8888
      - "8888:8888"
    depends_on:
      - influxdb
    networks:
      - senseops


networks:
  senseops:
    external:
      name: senseops

**Comment by sk0010637** [comment]
Hello,

Does this mean this, we will not be able to add any new container in the existing implementation of docker-compose.yml file unless we update the set up to the current version.

**Comment by sk0010637** [comment]
Hello ,

Request you to let me know if we can have a short connect regarding this.   This is basically Telia's grafana implementation where I am trying to add one more environment for server health monitoring through Grafana.

**Comment by mountaindude** [comment]
You can certainly have multiple Butler SOS instances running in parallel, but you must make sure that the ports exposed by each instance do not conflict with any other ports used on the server in question. This is standard server stuff and not a limitation in Butler SOS - it's just how microservices and APIs work.

I can't see why you could not run several instances of Butler 4.0.0 (which you seem to be using now) in parallel, but there have been significant improvements in the more recent versions, so I strongly suggest you move to the latest version.


I am personally on vacation and won't be able to support other than occasionally via this channel, I am afraid.

**Comment by mountaindude** [comment]
Adding to the above, you can in docker-compose.yml files specify that multiple containers should be started, for example one Butler SOS for a development Qlik Sense cluster, one Butler SOS for the Sense test cluster and one Butler SOS for the production cluster.

In that example you will need 3 different Butler SOS config files, each including the correct info for the cluster in question.
Certificates must be exported from the QMC of each cluster. 
Basically repeating the Butler SOS config process for each cluster, but also making sure that ports exposed, container names etc are unique across the started containers. 

This concept has been implemented lots of times, both using Docker containers and running Butler SOS as native Node.js apps. The Docker route is preferred though, as it usually gives better process isolation and stability.

**Comment by sk0010637** [comment]
Hello Goran,

Hope you are back from the vacation and enjoyed the same.  

Request you to please let me know how to get the latest version of docker-compose.yml file.

I am still getting the same error 

ERROR: The Compose file './docker-compose.yml' is invalid because:
Unsupported config option for services.butler-sos_FIN_dev: 'enableMQTT'
Unsupported config option for services.butler-sos_dev: 'enableMQTT'
Unsupported config option for services.butler-sos_prod: 'enableMQTT'

**Comment by sk0010637** [comment]
Hello Goran,

Hope you are back from the vacation and enjoyed the same.  

Request you to please let me know how to get the latest version of docker-compose.yml file.

I am still getting the same error 

ERROR: The Compose file './docker-compose.yml' is invalid because:
Unsupported config option for services.butler-sos_FIN_dev: 'enableMQTT'
Unsupported config option for services.butler-sos_dev: 'enableMQTT'
Unsupported config option for services.butler-sos_prod: 'enableMQTT'

**Comment by sk0010637** [comment]
Hello Goran,

Hope you are back from the vacation and enjoyed the same.  

Request you to please let me know how to get the latest version of docker-compose.yml file.

I am still getting the same error 

ERROR: The Compose file './docker-compose.yml' is invalid because:
Unsupported config option for services.butler-sos_FIN_dev: 'enableMQTT'
Unsupported config option for services.butler-sos_dev: 'enableMQTT'
Unsupported config option for services.butler-sos_prod: 'enableMQTT'

**Comment by sk0010637** [comment]
[docker-compose.yml.txt](https://github.com/ptarmiganlabs/butler-sos/files/5097677/docker-compose.yml.txt)

Hello Goran,

I have attached the docker-compose.yml file for your reference.

While trying to add new container (.butler-sos_FIN_dev), I am getting below error

I am still getting the same error

ERROR: The Compose file './docker-compose.yml' is invalid because:
Unsupported config option for services.butler-sos_FIN_dev: 'enableMQTT'
Unsupported config option for services.butler-sos_dev: 'enableMQTT'
Unsupported config option for services.butler-sos_prod: 'enableMQTT'

Request you to please let me know how to get the latest version of docker-compose.yml file.

Also, suggest for the error message.  It started giving the error of old container implementation as well.
(butler-sos_dev, butler-sos_prod)

**Comment by mountaindude** [comment]
I have removed the file you attached. Not a good idea from security perspective to attach files with sensitive info...

There is no "most recent version" of the `docker-compose.yaml` file. You just have to customise it to your needs. Every setup is unique when it comes to host names, where things are stored on disk etc etc. 
You must understand how Docker and docker-compose works if you are to use them.

Otherwise you can always run Butler SOS as a native Node.js application on a Linux or Windows server - works too. But then you of course need to know how to install services on those platforms. 

The "enableMQTT" error indicate that the configuration YAML file is incorrectly structures/formatted, in some way. 
I can only recommend that you look at https://butler-sos.ptarmiganlabs.com/ for info on configuration, and review your config files to make sure they are correctly formatted.
Please note that the documentation site contains info about the most recent version (which is not ver 4). I.e. if you follow the instructions on the doc site, you should make sure to upgrade to most recent version of Butler SOS too.

**Comment by mountaindude** [comment]
Oh, I just noted something:

In the docker-compose file that you attached (=the one that I subsequently deleted for security reasons) you have a couple of lines:

```
    mqttConfig:
    enableMQTT: false
```

Those lines don't belong in a docker-compose file, at least not in a basis setup as this after all is. 
Instead, those lines are part of the YAML config file (usually called production.yaml), which tells Butler SOS how to operate.

docker-compose docs [here](https://docs.docker.com/compose/).

**Comment by mountaindude** [comment]
Cleaning up old tickets - closing.

Please re-open if still an issue, or create a new ticket if new issues occur.
